packages/core/src/execution-engine/__tests__/execution-lifecycle-hooks.test.ts

```diff
@@ -0,0 +1,107 @@
+
+	let hooks: ExecutionLifecycleHooks;
+	beforeEach(() => {
+		jest.clearAllMocks();
+		hooks = new ExecutionLifecycleHooks('internal', executionId, workflowData);
+	});
+
+	describe('constructor()', () => {
+		it('should initialize with correct properties', () => {
+			expect(hooks.mode).toBe('internal');
+			expect(hooks.executionId).toBe(executionId);
+			expect(hooks.workflowData).toBe(workflowData);
+			expect(hooks.handlers).toEqual({
+				nodeExecuteAfter: [],
+				nodeExecuteBefore: [],
+				nodeFetchedData: [],
+				sendResponse: [],
+				workflowExecuteAfter: [],
+				workflowExecuteBefore: [],
+			});
+		});
+	});
+
+	describe('addHandler()', () => {
+		const hooksHandler =
+			mock<{
+				[K in keyof ExecutionLifecyleHookHandlers]: ExecutionLifecyleHookHandlers[K][number];
+			}>();
+
+		const testCases: Array<{ hook: ExecutionLifecycleHookName; args: unknown[] }> = [
+			{ hook: 'nodeExecuteBefore', args: ['testNode'] },
+			{
+				hook: 'nodeExecuteAfter',
+				args: ['testNode', mock<ITaskData>(), mock<IRunExecutionData>()],
+			},
+			{ hook: 'workflowExecuteBefore', args: [mock<Workflow>(), mock<IRunExecutionData>()] },
+			{ hook: 'workflowExecuteAfter', args: [mock<IRun>(), mock<IDataObject>()] },
+			{ hook: 'sendResponse', args: [mock<IExecuteResponsePromiseData>()] },
+			{ hook: 'nodeFetchedData', args: ['workflow123', mock<INode>()] },
+		];
+
+		test.each(testCases)('should add and process $hook hooks', async ({ hook, args }) => {
+			hooks.addHandler(hook, hooksHandler[hook]);
+			await hooks.runHook(hook, args);
+			expect(hooksHandler[hook]).toHaveBeenCalledWith(...args);
+		});
+	});
+
+	describe('runHook()', () => {
+		it('should execute multiple hooks in order', async () => {
+			const executionOrder: string[] = [];
+			const hook1 = jest.fn().mockImplementation(async () => {
+				executionOrder.push('hook1');
+			});
+			const hook2 = jest.fn().mockImplementation(async () => {
+				executionOrder.push('hook2');
+			});
+
+			hooks.addHandler('nodeExecuteBefore', hook1, hook2);
+			await hooks.runHook('nodeExecuteBefore', ['testNode']);
+
+			expect(executionOrder).toEqual(['hook1', 'hook2']);
+			expect(hook1).toHaveBeenCalled();
+			expect(hook2).toHaveBeenCalled();
+		});
+
+		it('should maintain correct "this" context', async () => {
+			const hook = jest.fn().mockImplementation(async function (this: ExecutionLifecycleHooks) {
+				expect(this.executionId).toBe(executionId);
+				expect(this.mode).toBe('internal');
+			});
+
+			hooks.addHandler('nodeExecuteBefore', hook);
+			await hooks.runHook('nodeExecuteBefore', ['testNode']);
+
+			expect(hook).toHaveBeenCalled();
+		});
+
+		it('should handle errors in hooks', async () => {
+			const errorHook = jest.fn().mockRejectedValue(new Error('Hook failed'));
+			hooks.addHandler('nodeExecuteBefore', errorHook);
+
+			await expect(hooks.runHook('nodeExecuteBefore', ['testNode'])).rejects.toThrow('Hook failed');
+		});
+	});
+});
```

Comment:
More ideas :)
- if a hook fails, should continue executing remaining hooks 
- if a hook has no handlers, running the hook should not throw
- if two handlers are added for a hook, both should be called (not overridden)